modules/ghlint-rules/src/main/kotlin/net/twisterrob/ghlint/rules/RequiredPermissionsRule.kt

```diff
@@ -0,0 +1,229 @@
+package net.twisterrob.ghlint.rules
+
+import net.twisterrob.ghlint.model.Access
+import net.twisterrob.ghlint.model.Permission
+import net.twisterrob.ghlint.model.Scope
+import net.twisterrob.ghlint.model.WorkflowStep
+import net.twisterrob.ghlint.model.effectivePermissions
+import net.twisterrob.ghlint.model.effectiveScopes
+import net.twisterrob.ghlint.rule.Example
+import net.twisterrob.ghlint.rule.Issue
+import net.twisterrob.ghlint.rule.Reporting
+import net.twisterrob.ghlint.rule.report
+import net.twisterrob.ghlint.rule.visitor.VisitorRule
+import net.twisterrob.ghlint.rule.visitor.WorkflowVisitor
+
+private typealias InferRequiredPermissions = (step: WorkflowStep.Uses) -> Set<RequiredScopes>
```

Comment:
@TWiStErRob Why is this a Set<RequiredScopes>? RequiredScopes contains a Set?

Reply:
I think we should model that structure inside the RequiredScopes if we want to include a reason for every set of scopes, otherwise this is getting quite high brain complexity to parse now

Reply:
```kotlin
private typealias InferRequiredPermissions = (step: WorkflowStep.Uses) -> Set<RequiredScope>

private data class RequiredScope(
	val scope: Scope,
	val reason: String,
)
```

This makes more sense to me?

Reply:
https://github.com/TWiStErRob/net.twisterrob.ghlint/commit/33a8fe9667401e3425393c14a8fb6a9741e23c69

Tests passing with this - happy to commit it if you like it?

Reply:
The code you commented out shows why it needs to be a set inside a set, that implementation is correct (no tests though 😅). There might be multiple permissions that play together, otherwise we would need to duplicate the reason for each permission.

This duplication might not be a bad thing, actually, because it forces us to be explicit why it's needed without potentially being handwavy. However it might generate "noise" when there are no permissions declared, 3 steps needing a total of 10 permissions for 4 reasons would report 10 findings, even though there are only 4 would be necessary.

(This brings up an interesting question, should this rule be disabled when there are no `permissions:` present? Not because it's undecidable from repo settings, but because the user might've opted to not define permissions. Although at that point they could also suppress this rule for that use case / after suppression is possible of course)

Go ahead, commit, but make the 2nd example compile and make sense too. I intentionally picked a more complex one to force use to design more flexibly.